### PR TITLE
CA2208 Instantiate argument exceptions correctly

### PR DESCRIPTION
**Pull request by elachlan**
Relates to #7174
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/CA2208

### CODE REVIEW

**Review by Forgind — APPROVED** [review]

File: src/Build.OM.UnitTests/TransientIO.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ public string GetAbsolutePath(string relative)
             var absolute = Path.GetFullPath(Path.IsPathRooted(relative) ? relative : Path.Combine(tempRoot, relative));
             if (!IsControled(absolute))
             {
-                throw new ArgumentException(nameof(relative));
+                throw new ArgumentException(null, nameof(relative));
```
Comment (sharwell) [inline src/Build.OM.UnitTests/TransientIO.cs:80]: 💭 I would prefer to see a better message here if possible

Comment (elachlan) [reply src/Build.OM.UnitTests/TransientIO.cs:80]: It looks like the code is checking that the relative path is indeed relative to the RootFolder. So I supposed the error should be something like:
`The path is invalid. Either the path is an absolute path that is not rooted to the RootFolder, or just an invalid path`.

Comment (Forgind) [reply src/Build.OM.UnitTests/TransientIO.cs:80]: `Either the path is invalid or it is not rooted to RootFolder.`
? Bit more concise.

**Comment by elachlan** [comment]
Build is failing on dependency microsoft.codeanalysis.collections:
```
/home/vsts/.nuget/packages/microsoft.codeanalysis.collections/4.0.0-4.21379.20/contentFiles/cs/netstandard2.0/ImmutableSegmentedList`1.cs(346,23): error CA2208: (NETCORE_ENGINEERING_TELEMETRY=Build) Call the ArgumentException constructor that contains a message and/or paramName parameter
/home/vsts/.nuget/packages/microsoft.codeanalysis.collections/4.0.0-4.21379.20/contentFiles/cs/netstandard2.0/ImmutableSegmentedList`1+ValueBuilder.cs(249,27): error CA2208: (NETCORE_ENGINEERING_TELEMETRY=Build) Call the ArgumentException constructor that contains a message and/or paramName parameter
```

**Comment by elachlan** [comment]
Related Roslyn issue:
https://github.com/dotnet/roslyn/issues/45286

**Comment by Forgind** [comment]
So we're waiting for that issue to be resolved? That looks outside the MSBuild layer.

**Comment by elachlan** [comment]
Yes, I am not sure how to tell the analyzer not to analyze code like this. So I am attempting to fix the upstream code.

**Comment by Youssef1313** [comment]
@elachlan Those errors should be suppressible via editorconfig per https://github.com/dotnet/roslyn/issues/55992.

If it doesn't work (I haven't tested), then dotnet/roslyn#55992 should probably re-considered.

FYI @jaredpar

**Comment by elachlan** [comment]
.editorconfig didn't work. But I don't think we are using IncludeContentInPack.

**Comment by Youssef1313** [comment]
@elachlan I can't think of how `Microsoft.CodeAnalysis.Collections` contents are seen by the compiler without, somehow, `IncludeContentInPack` being set (on Roslyn side I mean).

Since you can confirm that `.editorconfig` doesn't work. I'd like @jaredpar to re-consider something for dotnet/roslyn#55992

**Comment by elachlan** [comment]
See the commit adding it here:
https://github.com/dotnet/msbuild/commit/f1cd160dbd39718ac7bade9cdd1f1a84eef40493

**Comment by Youssef1313** [comment]
@elachlan I'm referring to roslyn side, the package itself has its source code included in pack. That's why I think dotnet/roslyn#55992 is relevant.

**Comment by elachlan** [comment]
@sharwell in https://github.com/dotnet/roslyn/issues/55992 it was pointed out that globalconfig defined analyzers apply to source code supplied by nuget packages as well. Where as .editorconfig files do not apply analyzers to against the packages.

This means I will unfortunately need to move back to using .editorconfigs. Unless you know of a better way?

**Pull request closed without merging**